### PR TITLE
[top] Pinout and chip-level synthesis updates

### PR DESCRIPTION
**Pull request by msfschaffner**
This draft PR contains several changes that carry over the chip-level synthesis scripts from Bronze, and the planned pinout (both these things are intertwined due to the I/O SDC constraints).

There are still several things that need to be fixed before this PR is ready for review:
- [x] A dummy `spi_host` design has to be added for integration purposes, see #5178
- [x] The UART pin mapping needs to be fixed for FPGA emulation since the UART has become an MIO now (instead of a DIO)
- [x] The chip-level simulation environment needs to be adjusted in order to reflect the updated pinout.

I'll ping this PR again once I am done with the opens.

Note that this is only a preparatory step for the actual pinmux/padctrl overhaul.
There are still several Bronze items I will need to clean up after this PR went in.
However, putting the pinout in place now (even if it is in a slightly Bronzish way) allows us to deliver an early Silver pre-release to Nuvoton to get PD and DFT insertion going.

### CODE REVIEW

**Comment by msfschaffner** [comment]
Note that this is currently in a messy state. 
I'll ping you again when this has become stable.

**Comment by eunchan** [comment]
Good to have SPI change here :) I was working on it but the changes look same as mine. Thanks!

**Comment by msfschaffner** [comment]
> Good to have SPI change here :) I was working on it but the changes look same as mine. Thanks!

Great, thanks for giving this a look @eunchan!

File: hw/top_earlgrey/dv/tb/tb.sv
Comment (msfschaffner) [inline hw/top_earlgrey/dv/tb/tb.sv:65]: @weicaiyang @sriyerg @cindychip 

We are carrying over the Bronze pinout to master to enable Nuvoton to work on a first release candidate for Silver.

Can you take a look at this commit and let me know if you are ok with these changes?
Most of them should really just be temporary in order to get an initial version of the pinout merged.
We will have to iterate on this a couple of times when the templated pinmux changes and final pinout land.
Also, I will have to remove the JTAG mux and disentangle SPI and JTAG at some point.

**Comment by msfschaffner** [comment]
Last force push rebases this onto the freshly merged pinmux updates, and a couple of other changes on master.
Still working on FPGA and DV alignments...

**Comment by msfschaffner** [comment]
CI should pass now with the latest changes.

@vogelpi, can you do me a favor and check whether englishbreakfast still works after this change?

PS: I am going to squash some of the commits once this is approved.

File: sw/device/lib/pinmux.c
Comment (msfschaffner) [inline sw/device/lib/pinmux.c:14]: @silvestrst @GregAC 

Since the UART is now a muxed peripheral, I had to hardcode the corresponding pinmux configuration here to get our simulation and FPGA emulation environments up and running again.

This is not pretty, and intended to be a temporary workaround until the pinmux/padctrl/pinout changes went in and until the DIF is available.

Let me know if you think this should be done differently.

**Comment by msfschaffner** [comment]
@dsandrs FYI: this one needs to land before we can do a pre-release.

File: sw/device/lib/pinmux.c
Comment (silvestrst) [reply sw/device/lib/pinmux.c:14]: @msfschaffner thank you, looks good to me!

I think eventually the plan is to remove this file, and to the configuration directly in the bootrom using `dif_pinmux`.

File: hw/top_earlgrey/rtl/autogen/top_earlgrey_pkg.sv
Comment (vogelpi) [inline hw/top_earlgrey/rtl/autogen/top_earlgrey_pkg.sv:426]: Nit: Indentation

**Review by vogelpi — COMMENTED** [review]
Hey @msfschaffner I did a first quick review of the English Breakfast stuff and it looks good. But to be sure, I have to test it on the FPGA. I will get back to you.

**Comment by vogelpi** [comment]
I've tested this on the CW305 FPGA board (top_englishbreakfast). It doesn't work at the moment. I get the boot ROM output and can transfer a binary over SPI, but then it stops. I need to do some more debugging. Anyway, it would be nice if you could rebase this PR on top of lowRISC/OpenTitan#5255 (which does some required re-alignment). Doing the rebase will lead to some confilcts but they are easy to solve.

File: hw/top_englishbreakfast/data/top_englishbreakfast.hjson
Comment (vogelpi) [inline hw/top_englishbreakfast/data/top_englishbreakfast.hjson:666]: Note that when doing the rebase on top of #5255, you'll have to re-add `uart0` in the `mio` section.

File: sw/device/lib/pinmux.c
Comment (vogelpi) [reply sw/device/lib/pinmux.c:20]: Follow-up question: I can see where 32 and 33 come from. But out of curiosity, where does the 42 come from? Is there a big, earth-sized machine somewhere that spit this out? ;-)

**Review by vogelpi — COMMENTED** [review]
Hey @msfschaffner,
did some more debugging on the CW305 and it seems I can't get any UART data through the pinmux. I get the UART output on a separate/unmuxed debug pin though. So, I was wondering if the pinmux setup is really correct.

File: hw/top_earlgrey/rtl/autogen/top_earlgrey_pkg.sv
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/autogen/top_earlgrey_pkg.sv:426]: ah yeah right, will fix the template.

File: sw/device/lib/pinmux.c
Comment (msfschaffner) [reply sw/device/lib/pinmux.c:20]: yes, the earth sized machine is called `topgen` and the index itself is a function of the definition order inside the `mio_modules` list in `top_*.hjson` and the order inside the IP-specific available IO lists.

it looks like this is also the cause for the incorrect pinmux config on the englishbreakfast top.
i.e.,  since the enumeration order of the `gpio` and `uart0` modules in the `mio_modules` list was not the same. 

I fixed the order now, and slightly changed topgen such that it keeps `inouts` and `inputs`/`outputs` in-order instead of stacking all `inouts` first.

File: hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv
Comment (msfschaffner) [reply hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv:166]: yes that should be `10'bz`, will fix

File: hw/top_englishbreakfast/data/top_englishbreakfast.hjson
Comment (msfschaffner) [reply hw/top_englishbreakfast/data/top_englishbreakfast.hjson]: jep, done

**Comment by msfschaffner** [comment]
@eunchan @tjaychen can you take a look as well?

File: hw/top_earlgrey/data/top_earlgrey.sv.tpl
Comment (tjaychen) [inline hw/top_earlgrey/data/top_earlgrey.sv.tpl:739]: i'm curious, why not just `sig[width]-1`? is it because you're trying to place the highest indices first?

Comment (tjaychen) [reply hw/top_earlgrey/data/top_earlgrey.sv.tpl:739]: definitely something for another PR..but i wonder if we should wrap these up in separate python functions eventually.. just because so much of the code is repeated, with only the text being different.

It might make it even HARDER to read though..so i'm not super convinced.

File: hw/top_earlgrey/syn/asic.constraints.sdc
Comment (tjaychen) [inline hw/top_earlgrey/syn/asic.constraints.sdc:19]: this might be corrected in later commits already..but `ast_wrapper` doesn't exist anymore.

File: hw/ip/pinmux/rtl/jtag_mux.sv
Comment (tjaychen) [inline hw/ip/pinmux/rtl/jtag_mux.sv:92]: btw @msfschaffner completely separate topic... i kind of didn't realize at the time pinmux does native debouncing already.. maybe it would make sense to re-use the wakeup logic / debounce for usb wake.  But I would probably still keep the usb wake module to handle other wakeup related things.  

Anyways...something to do later :) I think we'd still need the index below though..because we'd still need to know which wakeup to feed in..minor optimizations I guess.

File: hw/ip/pinmux/rtl/padring.sv
Comment (tjaychen) [inline hw/ip/pinmux/rtl/padring.sv:65]: we're going to have to figure out how to properly handle this one long term..
technically there is no dedicated clock pad.. the external clock is only muxed.

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (tjaychen) [inline hw/top_earlgrey/rtl/top_earlgrey_asic.sv:8]: ah okay this comment explains whatever was above.
I think ast's internal model already does not need the external clock.  So I think the clock was mostly for consistency with the FPGA version.

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (tjaychen) [inline hw/top_earlgrey/rtl/top_earlgrey_asic.sv:105]: since usb is a dedicated module right now, do these at the moment need to automatically become dedicated signal also?
unless we make a topgen change?

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (tjaychen) [inline hw/top_earlgrey/rtl/top_earlgrey_asic.sv:119]: do we plan to generate this long term?

File: hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv
Comment (tjaychen) [inline hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv:101]: btw what happens to IOC12? are we going to add it later?

File: hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv
Comment (tjaychen) [inline hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv:86]: ah, i just remembered we decided to make at least one of the rbox signals dedicated.  Issue for another time? Or I guess you need the `rbox` integrated first?

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (tjaychen) [inline hw/top_earlgrey/rtl/top_earlgrey_asic.sv:181]: this is a dumb question...but why are all these dedicated pad variants unused..? not sure what i'm missing.

**Review by tjaychen — APPROVED** [review]
this lgtm with some clarifying comments and questions.
I don't expect all comments to be addressed, although we should probably track the pending items if they are indeed pending items.

**Comment by tjaychen** [comment]
@msfschaffner do you need any help tracking down some of the FPGA / simulation errors? just let me know if you want a second pair of eyes.

File: hw/top_earlgrey/data/top_earlgrey.sv.tpl
Comment (msfschaffner) [reply hw/top_earlgrey/data/top_earlgrey.sv.tpl:739]: yeah this needs to be in descending order.

we should definitely start to factor out some of these things into functions.
this is mostly a carry-over from Bronze to get things going.

File: hw/top_earlgrey/syn/asic.constraints.sdc
Comment (msfschaffner) [reply hw/top_earlgrey/syn/asic.constraints.sdc:19]: yeah that's right, this is again just carried over 1:1 from Bronze. 
several of these SDC constraints need to be reworked.

File: hw/ip/pinmux/rtl/jtag_mux.sv
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/jtag_mux.sv:92]: Makes sense to look into that, just didn't get to that yet.

File: hw/ip/pinmux/rtl/padring.sv
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/padring.sv:65]: Yeah I know. The pinout here is not entirely final, but just reflects what we had on Bronze.
Aligning these details is going to be a separate operation.

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/top_earlgrey_asic.sv:8]: Yeah that is one of the issues I need to solve. The FPGA tops currently instantiate the same padring instance, which is not ideal.
I'm thinking about making the padring either more parametric or even making the padring completely templated, and generate a different one for each top-level.

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/top_earlgrey_asic.sv:181]: these are the USB DIO signals that are either entirely unused, or that are currently connected to the USB differential pad instance, which is separate from the padring.

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/top_earlgrey_asic.sv:119]: Yeah this is just the Bronze workaround. 
The plan is to generate all this stuff from an updated pinmux config datastructure in the Hjson.

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/top_earlgrey_asic.sv:105]: Yeah the MIO/DIO distinction is pretty binary at the moment in topgen.
Also, not all top-levels use these signals in the same way.
As you see here, some of these signals are tied off since they are unused in certain tops.
This is also the reason why some of the DIO pads are marked as unused in the padring.

A fully templated padring could circumvent this problem, but we would have to generate one for each of the tops (i.e., FPGA targets, Verilator, ASIC).

File: hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv
Comment (msfschaffner) [reply hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv:101]: Yeah going to add later. As I mentioned, this just carries over stuff from Bronze to get Nuvoton started, and will have to undergo some additional alignments later on.

File: hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv
Comment (msfschaffner) [reply hw/top_englishbreakfast/rtl/top_englishbreakfast_cw305.sv:86]: Ideally rbox would be integrated for that, yes, because otherwise we would have to add an empty shell for that.
In the meantime, I plan to work on an updated pinmux/pinout configuration scheme. 
We will for instance also need the ability to make the MIO/DIO decision on finer granularity, and I plan to support that with the updated configuration scheme.

See also opens here:
https://github.com/lowRISC/opentitan/issues/5221

**Comment by msfschaffner** [comment]
Hey @tjaychen, thanks giving this a thorough look.
Most of these pending items are part of the planned pinmux/padctrl/padring overhaul which I started to track here: https://github.com/lowRISC/opentitan/issues/5221
This is PR is mainly carrying over the current Bronze configuration such that we can get Nuvoton started, and will need further alignments.

As for the CI failures - I'm about to take another look. 
I had them all debugged and got them passing, but the latest rebase onto the current master seems to have introduced new issues...

**Comment by msfschaffner** [comment]
Looks like I found the culprit: I used the wrong UART peripheral index in `pinmux.c`. It should have selected the 33rd peripheral input and output, for the UART RX/TX pins, and I accidentally used the value `33` in the code instead of `33-1=32`.

**Comment by msfschaffner** [comment]
The last force push aligns the SDC constraints with the current state on master.

File: hw/top_earlgrey/syn/asic.constraints.sdc
Comment (msfschaffner) [reply hw/top_earlgrey/syn/asic.constraints.sdc]: I have aligned all these SDC constraints such that they work with the current state of the design.

File: sw/device/lib/pinmux.c
Comment (vogelpi) [reply sw/device/lib/pinmux.c]: Ah I see, thanks for clarifying and fixing/stabilizing the issue! One day `topgen` will conquer the world. Could it be the OT is just about coming up with `topgen`?

**Review by vogelpi — APPROVED** [review]
Thanks @msfschaffner for doing the rebase and fixing the ordering of the inouts/inputs/outputs in topgen. I successfully tested this PR on CW305 (top_englishbreakfast), LGTM!

**Comment by msfschaffner** [comment]
Thanks @vogelpi! The last commit rebases everything on master and squashes several commits together.\
I'll merge this once CI is green.

**Pull request merged**
